### PR TITLE
[JBPM-9891] Outdated TaskInstanceView

### PR DESCRIPTION
**Pull request by fjtirado**
Using LinkedHashMap rather than LinkedHashSet to allow updating already
existing view instances, so the event emitter reflects the up-to-date
state of the entity represented by the view.
when it is committed in case several operations over that entity are
performed within the same transaction

**JIRA**:  

[link](https://issues.redhat.com/browse/JBPM-9891)

### CODE REVIEW

File: jbpm-persistence/jbpm-persistence-api/src/main/java/org/jbpm/persistence/api/integration/base/BaseEventCollection.java [hunk 1]
```diff
@@ -16,41 +16,61 @@
...
     }
-    
+
     @Override
     public void add(InstanceView<?> event) {
-        this.events.add(event);
+        this.eventsMap.put(event, event);
+    }
+
+    private void readObject(ObjectInputStream input) throws ClassNotFoundException, IOException {
```
Comment (fjtirado) [inline jbpm-persistence/jbpm-persistence-api/src/main/java/org/jbpm/persistence/api/integration/base/BaseEventCollection.java:62]: This is added for backward compatibility with pre-existing streams

**Comment by fjtirado** [comment]
jenkins retest this

**Review by gmunozfe — COMMENTED** [review]
Looks good to me, bang-up work @fjtirado !
Just a question when reinitializing eventsMap inside readObject method.
Is it possible to create a test for assuring this fix is working inside a transaction, and the TaskInstanceView keeps updated if modified twice?

**Comment by elguardian** [comment]
to be honest I don't see the point of switching the collection type. Set should be the right choice. replacing an element in a set is a matter of remove and add.
In any case whatever @gmunozfe accepts it or not. Please add the test.

**Comment by fjtirado** [comment]
> to be honest I don't see the point of switching the collection type. Set should be the right choice. replacing an element in a set is a matter of remove and add.
> In any case whatever @gmunozfe accepts it or not. Please add the test.
Im trying to avoid contains, remove and add sequence, which from performace point of view is pretty bad (internally the set is just a capped map). Trust me, this way is far better ;)

**Comment by fjtirado** [comment]
Ill add test for this case

File: jbpm-persistence/jbpm-persistence-api/src/main/java/org/jbpm/persistence/api/integration/base/BaseEventCollection.java [hunk 2]
```diff
@@ -16,41 +16,61 @@
...
+        input.defaultReadObject();
+        if (events != null && !events.isEmpty()) {
+            eventsMap = new LinkedHashMap<>();
```
[the remark this answers is not in the document]
Comment (fjtirado) [reply jbpm-persistence/jbpm-persistence-api/src/main/java/org/jbpm/persistence/api/integration/base/BaseEventCollection.java:65]: Added two test, one for deserialization and another for checking that if two object which equals return true are added to the collection, the collection size is 1, but the value stored is the latter inserted

**Comment by fjtirado** [comment]
@elguardian @gmunozfe Test added

**Comment by gmunozfe** [comment]
jenkins retest this

**Review by gmunozfe — APPROVED** [review]
Looks good to me, thanks for adding the tests, great work @fjtirado !

**Pull request merged**